### PR TITLE
Feature/issue 492 promotor fun refactor

### PR DESCRIPTION
**Pull request by mitzimorris**
#### Submisison Checklist

- [x] Run unit tests: `./runTests.py test/unit`
- [x] Run cpplint: `make cpplint`
- [x] Declare copyright holder and open-source license: see below

#### Summary:
Added new template specialization to stan/math/prim/mat/array_builder.hpp needed by stan lang generator for matrix expressions.  Refactored stan::math::promoter class into scalar, array, and matrix specializations and added doxygen comments to the source code.  Wrote unit tests for both array_builder and promoter functions.

#### Intended Effect:
Adds functionality needed for Stan language compiler; code cleanup.

#### How to Verify:
Unit tests.

#### Side Effects:
None.

#### Documentation:
Doxygen comments in source code.

#### Reviewer Suggestions: 
Sean Talts or Bob Carpenter

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company):  Mitzi Morris

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

**Comment by bob-carpenter** [comment]
@seantalts Would you mind reviewing this one?

**Comment by seantalts** [comment]
Sure thing!

**Comment by seantalts** [comment]
Hey, things are looking good but I'm realizing I don't really understand the objective - can you tell me more about why `array_builder` needs a specialization for `array_builder<Matrix<T, R, C>`? From my naive reading it would have seemed like the `promoter<Eigen::Matrix<F, R, C>, Eigen::Matrix<T, R, C> >` specialization would have handled this case inside of `array_builder` without `array_builder` needing to become Matrix-aware. But clearly we needed it so I think there's probably a gap in my mental model of C++ templating...

Also, would you mind if I asked some questions about the underlying code that isn't really related to the pull request? For example, I know you didn't change this, but I'm curious why we have our own `stan::math::matrix_d` when there is an `Eigen::MatrixXd`?

**Comment by mitzimorris** [comment]
The above example doesn't line up with the way that the Stan language compiler works, because the parser composes a matrix expression as a vector of row vectors.
Given the statement:
```
  matrix[1,1] td_mat11 = [ [ 0 ], [ 1] ];
```
The parser parses the expression `[[ 0], [ 1]]` into a matrix expression node which contains a list of the expression contents, each of which is a row_vector expression node.
The generated `c++` code is for the assignment statement is:
```
stan::math::assign(td_mat12, stan::math::to_matrix(stan::math::array_builder<Eigen::Matrix<double, 1, Eigen::Dynamic> >().add(td_rv2).array()));
```

I don't know why we have `stan::math::matrix_d`.
Another question is why both `stan::math::promoter` and `stan::math::promote_to`?

**Comment by mitzimorris** [comment]
The array_builder template only needs to have parameter `<T>`; the speciallization for Eigen::Matrix types is:
```
template <typename T>
    struct array_builder<Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic> >
```

Your observation that we don't need parameters `R` and `C` is correct and I've simplified the code accordingly.

**Comment by bob-carpenter** [comment]
> Also, would you mind if I asked some questions about the underlying code that isn't really related to the pull request? For example, I know you didn't change this, but I'm curious why we have our own stan::math::matrix_d when there is an Eigen::MatrixXd?

Feel free to drop this into a discourse thread if it's not related to this pull request.

We could drop matrix_d --- it's there for symmetry with matrix_v, matrix_fv, and matrix_ffv.  I tried at one point to get rid of those, but the replacements are a mouthful and they're pretty widely used.  I think it'd probably be better to just go to our typedefs wherever there's a concrete specialization.

File: stan/math/prim/arr/fun/array_builder.hpp [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+      array_builder() : x_() { }
+      template <typename F>
+      array_builder& add(const F& u) {
+        T t;
+        promoter<F, T>::promote(u, t);
+        x_.push_back(t);
+        return *this;
+      }
+
+      std::vector<T> array() {
```
Comment (bob-carpenter) [inline stan/math/prim/arr/fun/array_builder.hpp]: The return can probably be `const std::vector<T>&` --- constant reference.  It'd mean the result only lasts as long as the array builder, but the client could also copy it out if necessary.

File: stan/math/prim/arr/fun/array_builder.hpp [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+     * in statements) using an argumentchaining add() method and
+     * a getter method array() to return the result.
+     *
+     * @tparam T1 type of arg1
+     */
+    template <typename T>
+    struct array_builder {
+      std::vector<T> x_;
+      array_builder() : x_() { }
+      template <typename F>
```
Comment (bob-carpenter) [inline stan/math/prim/arr/fun/array_builder.hpp]: Doc on all public functions, methods and constructors.  I think space between definitions would help rility.eadab

File: stan/math/prim/arr/fun/common_type.hpp [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+#include <vector>
+
+namespace stan {
+  namespace math {
+    /**
+     * Struct carries out type promotion.
+     * This specialization is for vectors of differing types.
+     *
+     * @tparam T1 type of arg1
+     * @tparam T2 type of arg2
```
Comment (bob-carpenter) [inline stan/math/prim/arr/fun/common_type.hpp]: I don't see any arguments---what are arg1 and arg2?

File: stan/math/prim/arr/fun/common_type.hpp [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+#define STAN_MATH_PRIM_ARR_FUN_COMMON_TYPE_HPP
+
+#include <stan/math/prim/scal/fun/common_type.hpp>
+#include <boost/math/tools/promotion.hpp>
+#include <vector>
+
+namespace stan {
+  namespace math {
+    /**
+     * Struct carries out type promotion.
```
Comment (bob-carpenter) [inline stan/math/prim/arr/fun/common_type.hpp]: This is too vague.

File: stan/math/prim/arr/fun/promoter.hpp [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+#include <vector>
+
+namespace stan {
+  namespace math {
+    /**
+     * Struct to hold static function for element-wise type promotion.
+     * This specialization is for vectors of differing types.
+     *
+     * @tparam F type of input element
+     * @tparam T type of output element
```
Comment (bob-carpenter) [inline stan/math/prim/arr/fun/promoter.hpp]: Is input/output same as arg1 and arg2 above?  And is one the result of promoting or something?  Input and output and arg1 and arg2 are hard to follow.

File: stan/math/prim/mat/fun/to_matrix.hpp [hunk 6]
```diff
@@ -70,6 +70,31 @@ namespace stan {
...
+     * with the same dimensions and indexing order.
+     *
+     * @tparam T type of the scalar
+     * @param vec Eigen vector of vectors of scalar values
+     * @return the matrix representation of the input
+     */
+    template <typename T>
+    inline Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic>
+    to_matrix(const std::vector< Eigen::Matrix<T, 1, Eigen::Dynamic> >& vec) {
+      size_t R = vec.size();
```
Comment (bob-carpenter) [inline stan/math/prim/mat/fun/to_matrix.hpp]: Don't use caps for local variables.

File: stan/math/prim/mat/fun/to_matrix.hpp [hunk 7]
```diff
@@ -70,6 +70,31 @@ namespace stan {
...
+      if (R != 0) {
+        size_t C = vec[0].size();
+        Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic> result(R, C);
+        T* datap = result.data();
```
Comment (bob-carpenter) [inline stan/math/prim/mat/fun/to_matrix.hpp]: This is much simpler.

```
for (int i = 0; i < result.rows(); ++i)
  result.row(i) = vec[i];
```

But I don't think that arg should be called `vec` --- just use a dummy variable like `x` instead of an abbreviated typename.

File: stan/math/prim/mat/fun/to_matrix.hpp [hunk 8]
```diff
@@ -70,6 +70,31 @@ namespace stan {
...
+     * Returns a matrix representation of the Eigen vector of standard vectors
+     * with the same dimensions and indexing order.
+     *
+     * @tparam T type of the scalar
+     * @param vec Eigen vector of vectors of scalar values
+     * @return the matrix representation of the input
+     */
+    template <typename T>
+    inline Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic>
+    to_matrix(const std::vector< Eigen::Matrix<T, 1, Eigen::Dynamic> >& vec) {
```
Comment (bob-carpenter) [inline stan/math/prim/mat/fun/to_matrix.hpp]: Don't need space before `Eigen`

File: stan/math/prim/scal/fun/common_type.hpp [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+#include <boost/math/tools/promotion.hpp>
+
+namespace stan {
+  namespace math {
+    /**
+     * Struct carries out type promotion.
+     *
+     * @tparam T1 type of arg1
+     * @tparam T2 type of arg2
+     */
```
Comment (bob-carpenter) [inline stan/math/prim/scal/fun/common_type.hpp]: Same comment as above.  What's arg1 and arg2?

**Comment by seantalts** [comment]
Sorry, wasn't trying to point out that we didn't need `R` and `C`, but that it seems like maybe we only need your new specialization for `to_matrix` and that we might not need any new code for `array_builder`? I think the generated cpp you have above should correspond to this test (right?), which also works on `develop`:

```cpp
TEST(MathMatrix, arrayBuilderRowVector) {
  using std::vector;
  using stan::math::array_builder;
  using Eigen::Matrix;
  using Eigen::Dynamic;

  Matrix<int, 1, Dynamic> v1(4);
  v1 << 1, 2, 3, 4;

  Matrix<float, 1, Dynamic> v2(4);
  v2 << 1.1, 2.2, 3.3, 4.4;

  vector<Matrix<double, 1, Dynamic> > v = array_builder<Matrix<double, 1, Dynamic> >()
    .add(v1)
    .add(v2)
    .array();

  EXPECT_FLOAT_EQ(1, v[0](0));
  EXPECT_FLOAT_EQ(1.1, v[1](0));
  EXPECT_FLOAT_EQ(4, v[0](3));
  EXPECT_FLOAT_EQ(4.4, v[1](3));
}
```
So we might not need the new specialization for `array_builder<Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic> >`?

File: stan/math/prim/scal/fun/promoter.hpp [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+      }
+      inline static T promote_to(const F& u) {
+        return u;
+      }
+    };
+
+    /**
+     * Struct which holds static function for element-wise type promotion.
+     * This specialization is for scalars of the same type.
+     *
```
Comment (seantalts) [inline stan/math/prim/scal/fun/promoter.hpp]: I don't think you changed this, but why do we need a specialization for two things of the same type? I see this pattern in a couple of places and it seems to have an identical implementation to the one for two different types, so I'm confused about what is going on here / why we need this in general...

File: test/unit/math/prim/arr/fun/promoter_test.cpp [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+
+TEST(MathFunctionsArrPromoter,intVec2doubleVec) {
+  vector<int> from;
+  from.push_back(1);
+  from.push_back(2);
+  from.push_back(3);
+  promoter<vector<int>,vector<double> > p;
+  typedef BOOST_TYPEOF(p.promote_to(from)) result_t;
+  bool same = boost::is_same<vector<double>, result_t>::value;
+  EXPECT_TRUE(same);
```
Comment (seantalts) [inline test/unit/math/prim/arr/fun/promoter_test.cpp]: Can you make these 3 lines a utility function? Seems useful.

**Comment by mitzimorris** [comment]
I've added doc comments to `stan/math/prim/*/fun/common_type.hpp` and `stan/math/prim/mat/fun/promote_common.hpp` and unit tests for `promote_common.hpp`.

This answers all the feedback from review.

It might be possible to do more code cleanup.  The promote_common function is used by matrix mult/divide functions:
- stan/math/fwd/mat/fun/mdivide_left_ldlt.hpp
- stan/math/prim/mat/fun/mdivide_left.hpp
- stan/math/prim/mat/fun/mdivide_left_ldlt.hpp
- stan/math/prim/mat/fun/mdivide_left_spd.hpp
- stan/math/prim/mat/fun/mdivide_left_tri.hpp
- stan/math/prim/mat/fun/mdivide_right.hpp
- stan/math/prim/mat/fun/mdivide_right_tri_low.hpp

only `promote_common` uses `common_type`

File: stan/math/prim/arr/fun/array_builder.hpp [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+        x_.push_back(promote_elements<T, S>::promote(u));
+        return *this;
+      }
+
+      /**
+       * Getter method to return array itself.
+       *
+       * @returns std:vector<T> of composed array elements.
+       */
+      std::vector<T> array() {
```
Comment (bob-carpenter) [inline stan/math/prim/arr/fun/array_builder.hpp]: Should this be returning a constant reference?  As is, this is going to copy.

File: stan/math/prim/arr/fun/array_builder.hpp [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+     * in statements) using an argumentchaining add() method and
+     * a getter method array() to return the result.
+     * Array elements are held in std::vector of type T.
+     *
+     * @tparam T type of array elements
+     */
+    template <typename T>
+    struct array_builder {
+      std::vector<T> x_;
+      array_builder() : x_() { }
```
Comment (bob-carpenter) [inline stan/math/prim/arr/fun/array_builder.hpp]: Doc please.  You can make the x_ private, too.

File: stan/math/prim/arr/fun/promote_elements.hpp [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+     */
+    template <typename T>
+    struct promote_elements<std::vector<T>, std::vector<T> > {
+      /**
+       * Return input vector.
+       *
+       * @param u vector of type T
+       * @returns vector of type T
+       */
+      inline static std::vector<T> promote(const std::vector<T>& u) {
```
Comment (bob-carpenter) [inline stan/math/prim/arr/fun/promote_elements.hpp]: Could you return a constant reference here, too?  Otherwise, there's another needless copy.  I think that should work OK with any usage you make of the return type.

File: stan/math/prim/mat/fun/promote_elements.hpp [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+    struct promote_elements<Eigen::Matrix<T, R, C>,
+                            Eigen::Matrix<T, R, C> > {
+      /**
+       * Return input matrix.
+       *
+       * @param u matrix of type T
+       * @returns matrix of type T
+       */
+      inline static Eigen::Matrix<T, R, C>
+      promote(const Eigen::Matrix<T, R, C>& u) {
```
Comment (bob-carpenter) [inline stan/math/prim/mat/fun/promote_elements.hpp]: I think this can return const reference, too.

File: stan/math/prim/mat/fun/to_matrix.hpp [hunk 16]
```diff
@@ -14,54 +14,53 @@ namespace stan {
...
      * @tparam C number of columns
-     * @param matrix matrix
+     * @param x matrix
      * @return the matrix representation of the input
      */
     template <typename T, int R, int C>
     inline Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic>
-    to_matrix(const Eigen::Matrix<T, R, C>& matrix) {
-      return matrix;
+    to_matrix(const Eigen::Matrix<T, R, C>& x) {
```
Comment (bob-carpenter) [inline stan/math/prim/mat/fun/to_matrix.hpp]: const reference return?

File: stan/math/prim/mat/fun/to_matrix.hpp [hunk 17]
```diff
@@ -14,54 +14,53 @@ namespace stan {
...
-        T* datap = result.data();
-        for (size_t i=0, ij=0; i < C; i++)
-          for (size_t j=0; j < R; j++, ij++)
-            datap[ij] = vec[j][i];
+    to_matrix(const std::vector< std::vector<T> >& x) {
+      size_t rows = x.size();
+      if (rows != 0) {
+        size_t cols = x[0].size();
+        Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic> result(rows, cols);
+        for (size_t i=0, ij=0; i < cols; i++)
```
Comment (bob-carpenter) [inline stan/math/prim/mat/fun/to_matrix.hpp]: Use `++i` rather than `i++` --- compiler can probably deal here, but only use `x++` when you are going to use the value of `x` before the increment.

File: stan/math/prim/mat/fun/to_matrix.hpp [hunk 18]
```diff
@@ -70,22 +69,46 @@ namespace stan {
...
-     * @param vec vector of vectors of integer values
+     * @tparam T type of the scalar
+     * @param x Eigen vector of vectors of scalar values
      * @return the matrix representation of the input
      */
+    template <typename T>
+    inline Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic>
+    to_matrix(const std::vector<Eigen::Matrix<T, 1, Eigen::Dynamic> >& x) {
+      size_t rows = x.size();
+      if (rows != 0) {
```
Comment (bob-carpenter) [inline stan/math/prim/mat/fun/to_matrix.hpp]: Generally we want the unusual case to come first and return and the usual case to not be nested in an else.  This was a biggy in the Refactoring book and I think it makes code a lot cleaner and easier to follow.

File: stan/math/prim/mat/fun/to_matrix.hpp [hunk 19]
```diff
@@ -70,22 +69,46 @@ namespace stan {
...
-    to_matrix(const std::vector< std::vector<int> >& vec) {
-      size_t R = vec.size();
-      if (R != 0) {
-        size_t C = vec[0].size();
-        Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic> result(R, C);
-        double* datap = result.data();
-        for (size_t i=0, ij=0; i < C; i++)
-          for (size_t j=0; j < R; j++, ij++)
-            datap[ij] = vec[j][i];
+    to_matrix(const std::vector< std::vector<int> >& x) {
```
Comment (bob-carpenter) [inline stan/math/prim/mat/fun/to_matrix.hpp]: same comment as last one

File: stan/math/prim/scal/fun/common_type.hpp [hunk 20]
```diff
@@ -0,0 +1,24 @@
...
+#include <boost/math/tools/promotion.hpp>
+
+namespace stan {
+  namespace math {
+    /**
+     * Struct which calculates type promotion given two types.
+     *
+     * <p>This is the base implementation for scalar types.
+     *
+     * @tparam T1 scalar type, either int, double, or stan::math::var
```
Comment (bob-carpenter) [inline stan/math/prim/scal/fun/common_type.hpp]: Why the restriction on types?  Will this not work when we start using `fvar<T>` for forward mode?  When I said all types would be int, double or autodiff variables, I meant all the different autodiff variables, but that if an autodiff variable was used, they'd all be the same autodiff variables.  That is, you wouldn't get an `fvar<var>` and a `var`.

File: stan/math/prim/scal/fun/promote_elements.hpp [hunk 21]
```diff
@@ -0,0 +1,51 @@
...
+     */
+    template <typename T>
+    struct promote_elements<T, T> {
+      /**
+       * Return input element.
+       *
+       * @param u input of type T
+       * @returns input as type T
+       */
+      inline static T promote(const T& u) {
```
Comment (bob-carpenter) [inline stan/math/prim/scal/fun/promote_elements.hpp]: return a const reference here --- always return a const reference when you can rather than copying.

File: stan/math/prim/arr/fun/array_builder.hpp (hunk 12 above)
Comment (seantalts) [inline stan/math/prim/arr/fun/array_builder.hpp]: If that happens, that requires the array_builder to stick around in memory as long as you want to use its `array()`, right? Is there a way to suggest to the compiler that you are giving up the internal `x_` without copying it? Or can the compiler be smart enough to detect that the internal `x_` is never used again and avoid the copy with the current semantics?

**Comment by bob-carpenter** [comment]
You're looking for the C++11 move semantics.  For now, we have to return a copy of anything in local scope because references don't last outside of their scope as you note.

**Comment by bob-carpenter** [comment]
Sorry about that.  I just meant you should check them all.  

You can't return references to locally allocated objects.

**Comment by mitzimorris** [comment]
Bob, per above comment, which methods are you referring to?
for class `promote_elements`,  method `promote` can return a const reference when the arg type matches the return type, because nothing new is created, is this correct?

File: stan/math/prim/arr/fun/array_builder.hpp (hunk 12 above)
Comment (bob-carpenter) [inline stan/math/prim/arr/fun/array_builder.hpp]: We're talking about `array_builder` method `array()`.  The question is whether the `array_builder` instance 

* holds a reference to the returned object as a member variable;  if it's only local, you can't return a reference, and

* persists through the lifespan of the returned reference

You can always copy on return.  For instance, if you have a function

```
const foo& bar();
```

then you can copy on use with:

```
foo x = bar();
```

or you can just preserve the reference without copying as

```
const foo& x = bar();
```

In the latter case, the reference source must persist through the use of `x`.
And of course, you can always copy on return from the outside.  For example, if I have

Comment (mitzimorris) [inline stan/math/prim/arr/fun/array_builder.hpp]: the array_builder object holds a reference to the array object as a private member variable.

in the generated stan c++ code, an array_builder object is instantiated and the array() method is called as the 2nd argument to the stan::math::assign function:

stan program stmt:
```
 int td_arr_int_d1_1[3] = {1, 2, 3};
```

compiled to:
```
        validate_non_negative_index("td_arr_int_d1_1", "3", 3);
        td_arr_int_d1_1 = std::vector<int>(3,int(0));
        stan::math::fill(td_arr_int_d1_1, std::numeric_limits<int>::min());
        stan::math::assign(td_arr_int_d1_1,static_cast<std::vector<int> >(stan::math::array_builder<int >().add(1).add(2).add(3).array()));
```
does this constitute copy on use?

declaring the array() method:
```
const std::vector<T>& array();
```
and then compiling and running the test models in `stan/src/test/test-models/good/array-expr/validate_array_expr_primitives` and `validate_array_expr_containers`, which have data.R files checked in for running the test models, produced no compiler or runtime errors.
